### PR TITLE
Add test to ensure certain names are not imported from `typing_extensions`

### PR DESCRIPTION
**Pull request by AlexWaygood**
Marking this as draft for now as:

1. I'm not currently on a computer with Python installed.
2. The test might fail at first on some typeshed stubs.

Let's see what happens!

Cc. @Akuli, @sirosen

### CODE REVIEW

**Comment by AlexWaygood** [comment]
^That last failed run shows that it works! And that I need to add `ParamSpec` to the hardcoded list.

**Comment by sirosen** [comment]
This looks great to me! It will need updates whenever new things are added to typing_extensions, but that's not too often.

I notice that there's not a visit_Import alongside visit_ImportFrom. Does some other check look for things like this?
```python
import typing_extensions as te

class P(te.Protocol): ...
```

I'm happy to add another check to the visitor if not (after this, to avoid conflicts).

**Comment by JelleZijlstra** [comment]
I'd prefer if we hardcode the list of names that must be imported from typing. If a new contributor starts to use a typing-extensions name in the future that's not in the list yet, we're going to get a confusing error telling them to use `typiing`.

**Comment by Akuli** [comment]
> import typing_extensions as te

The script isn't intended to be perfect. It's fine if there is a corner case in syntax that nobody uses in typeshed.

**Comment by AlexWaygood** [comment]
> This looks great to me! It will need updates whenever new things are added to typing_extensions, but that's not too often.
> 
> I notice that there's not a visit_Import alongside visit_ImportFrom. Does some other check look for things like this?
> 
> ```python
> import typing_extensions as te
> 
> class P(te.Protocol): ...
> ```
> 
> I'm happy to add another check to the visitor if not (after this, to avoid conflicts).

Yah, I considered this, but figured that `typing_extensions` is just a long name — it's unlikely that people will do a straight import `typing_extensions`, and as @Akuli says, the script doesn't need to be perfect. No need to complicate things 😉

**Comment by AlexWaygood** [comment]
> I'd prefer if we hardcode the list of names that must be imported from typing. If a new contributor starts to use a typing-extensions name in the future that's not in the list yet, we're going to get a confusing error telling them to use `typiing`.

Makes sense, I'll make the change.

**Comment by AlexWaygood** [comment]
It looks like that passes CI, so I'm going to introduce some Bad Imports from `typing_extensions` to check the refactoring works.

**Comment by AlexWaygood** [comment]
Yup, that works.

File: tests/check_new_syntax.py [hunk 1]
```diff
@@ -12,6 +12,23 @@
 CONTEXT_MANAGER_ALIASES = {"ContextManager": "AbstractContextManager", "AsyncContextManager": "AbstractAsyncContextManager"}
 CONTEXTLIB_ALIAS_ALLOWLIST = frozenset({Path("stdlib/contextlib.pyi"), Path("stdlib/typing_extensions.pyi")})
 
+IMPORTED_FROM_TYPING_NOT_TYPING_EXTENSIONS = frozenset(
+    {"ClassVar", "Type", "NewType", "overload", "Text", "Protocol", "runtime_checkable"}
```
Comment (JelleZijlstra) [inline tests/check_new_syntax.py:16]:
```suggestion
    {"ClassVar", "Type", "NewType", "overload", "Text", "Protocol", "runtime_checkable", "NoReturn"}
```
I believe this one is safe too

Comment (AlexWaygood) [reply tests/check_new_syntax.py:16]:
```python
>>> import typing_extensions
>>> "NoReturn" in typing_extensions.__all__
False
>>> typing_extensions.NoReturn
typing.NoReturn
```
What's up with this?

Comment (hauntsaninja) [reply tests/check_new_syntax.py:16]: https://github.com/python/typing/blob/a2371460d184c96aab7a69acc47fd059f875e3b4/typing_extensions/src/typing_extensions.py#L91

Comment (hauntsaninja) [reply tests/check_new_syntax.py:16]: Seems like an issue, want to open a PR at python/typing? I guess not many people do `from typing_extensions import *` + it only really affects Python 3.5 to 3.6.2

Comment (AlexWaygood) [reply tests/check_new_syntax.py:16]: Meh, I kind of don't feel like opening a PR for a Python version that will be officially unsupported from next week :p

It kind of seems like it only affects one person: me, as I was using `__all__` as a handy list of the names were public in `typing_extensions`

**Pull request merged**